### PR TITLE
Integrate Asset-Sync Features

### PR DESCRIPTION
**Pull request by samvaity**
This PR adds the Test Proxy asset-sync feature to Java repo

### CODE REVIEW

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/TestProxyRecordPolicy.java [hunk 1]
```diff
@@ -74,6 +80,16 @@ public void startRecording(String recordFile) {
         addProxySanitization(this.sanitizers);
     }
 
+    private String getAssetJsonPath() {
+        Path rootPath;
+        try {
+            rootPath = Paths.get(System.getProperty("user.dir") + "/.." + "/.." + "/..").toRealPath();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (samvaity) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/policy/TestProxyRecordPolicy.java:88]: update to logger

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestProxyTestBase.java [hunk 2]
```diff
@@ -33,7 +32,7 @@ public TestProxyTestBase() {
     public static void setupTestProxy() {
         testMode = initializeTestMode();
         if (isTestProxyEnabled() && (testMode == TestMode.PLAYBACK || testMode == TestMode.RECORD)) {
-            testProxyManager = new TestProxyManager(TestUtils.getRepoRoot());
+            testProxyManager = new TestProxyManager();
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestProxyTestBase.java:35]: So now that we've moved the test recordings the project does that have the `session-records` folder anymore and so it fails to create a path for it.

But leaving that aside. 
I would have thought the x-recording-file to be = `"sdk/appconfiguration/azure-data-appconfiguration/src/test/resources/session-records/"name.json"`

and not
`"sdk/appconfiguration/azure-data-appconfiguration/target/test-classes/session-records/"name.json"`

is that assumption right?

[the remark this answers is not in the document]
Comment (scbedd) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestProxyTestBase.java:35]: It depends on which files were commit with the recording, right?

Let's say during `recording`, you provide path `sdk/appconfiguration/azure-data-appconfiguration/src/test/resources/session-records/name.json`. That is the path that the file will be written to.

Then, during `playback`, you provide `sdk/appconfiguration/azure-data-appconfiguration/target/test-classes/session-records/name.json`. That would obviously fail as the path that they were written to doesn't align.

**tl;dr** it's totally up to which path you as the user provide during record and playback. The proxy doesn't actually care.

Btw the `"name.json` looks really odd to me. I'd definitely update that to a consistent path without the `"` in middle when you have a chance.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java [hunk 3]
```diff
@@ -148,4 +165,13 @@ public boolean didTestRun() {
     void setTestIteration(Integer testIteration) {
         this.testIteration = testIteration;
     }
+
+    /**
+     * Returns a boolean indicating if the test recording should be externalized.
+     *
+     * @return Flag indicating if the test recording should be externalized.
+     */
+    public boolean skipExternalizeRecording() {
```
Comment (srnagar) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java:174]: isExternalizeRecordingSkipped()

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java [hunk 4]
```diff
@@ -42,12 +45,17 @@ public TestContextManager(Method testMethod, TestMode testMode) {
      * @param enableTestProxy True if the external test proxy is in use.
      * @param recordWithoutRequestBodyClassAnnotation flag indicating if {@code RecordWithoutRequestBody} annotation
      * present on test class.
+     * @param testClassPath the test class path
+     * @param skipExternalizeRecording flag indicating if {@code SkipExternalizeTestRecording} annotation
```
Comment (srnagar) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java:49]: In what scenarios do we want the tests to control whether the recording should be externalized or not?

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Annotation given to some tests to skip externalizing recording to enable running test validations locally.
+ * <p>This requires the test recordings to be present src/test/resources/session-records folder</p>
+ */
+@Retention(RUNTIME)
+@Target(ElementType.METHOD)
```
Comment (srnagar) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java:17]: We should support having this annotation on the entire test class.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+package com.azure.core.test.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Annotation given to some tests to skip externalizing recording to enable running test validations locally.
```
Comment (srnagar) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java:13]: I am not sure we need this annotation. Validating locally or using externalized recordings should be similar and shouldn't require tests to have this annotation.

Comment (samvaity) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java:13]: I was thinking if it might be needed when you are performing local validations similar to this test [here](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-test/src/test/java/com/azure/core/test/TestProxyTests.java#L271) reading local session-record files we should at least have an option.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java (hunk 5 above)
Comment (samvaity) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java:17]: Would we ever have scenarios for the entire test class to not be externalized?

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java [hunk 7]
```diff
@@ -42,12 +45,17 @@ public TestContextManager(Method testMethod, TestMode testMode) {
...
+     * present on test method.
      */
-    public TestContextManager(Method testMethod, TestMode testMode, boolean enableTestProxy, boolean recordWithoutRequestBodyClassAnnotation) {
+    public TestContextManager(Method testMethod, TestMode testMode, boolean enableTestProxy, boolean recordWithoutRequestBodyClassAnnotation, Path testClassPath,
+                              boolean skipExternalizeRecording) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/TestContextManager.java:53]: We should think about making a properties bag to pass to the constructor as we keep growing the options

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java (hunk 6 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-test/src/main/java/com/azure/core/test/annotation/SkipExternalizeTestRecording.java:13]: I think we should add special handling to `azure-core-test` for this alone rather than something generalized as other SDKs shouldn't be doing something like that.

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/RecordFilePayload.java [hunk 8]
```diff
@@ -16,12 +16,20 @@ public class RecordFilePayload {
...
+    @JsonProperty("x-recording-assets-file")
+    private String assetFile;
+
     /**
      * Creates an instance of {@link RecordFilePayload}.
      * @param recordingFile The partial path to the recording file.
+     * @param assetFile The path to asset file.
      */
-    public RecordFilePayload(String recordingFile) {
+    public RecordFilePayload(String recordingFile, String assetFile) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/models/RecordFilePayload.java:30]: Don't we need `JsonCreator` here? Or does this only get serialized? If this class is serialize only could we add `WRITE_ONLY` to the JSON properties

File: sdk/core/azure-core-test/src/main/java/com/azure/core/test/utils/TestUtils.java [hunk 9]
```diff
@@ -144,22 +150,25 @@ private TestUtils() {
...
+     *
+     * @param testClassPath the test class path
+     * @param resolveFolder the folder parent to resolve the path until
      * @return The {@link Path} to the root of the repo.
-     * @throws RuntimeException The eng folder could not be located.
+     * @throws RuntimeException The specified folder could not be located.
      */
-    public static Path getRepoRoot() {
-        Path path = getRecordFolder().toPath();
+    public static Path getRepoRootResolveUntil(Path testClassPath, String resolveFolder) {
```
Comment (alzimmermsft) [inline sdk/core/azure-core-test/src/main/java/com/azure/core/test/utils/TestUtils.java:160]: We should add some logic to prevent file walking from escaping outside of the azure-sdk-for-java root.

**Review by alzimmermsft — APPROVED** [review]

**Review by srnagar — APPROVED** [review]
